### DISCUSSION TITLE
What would be the hurdles for Android support with CPU devices

### DISCUSSION THREAD

**Discussion by widiba03304** [pocl/pocl#1632] [Q&A] [answered]
Hi,

I’m interested in building this project for Android devices, but it appears that PoCL has dropped support for Android with CPU devices, as noted in `doc/sphinx/source/android.rst`:

> A number of features in PoCL such as CPU devices and the icd loader are not available on Android. Below is a list of recommended CMake options::
> -DENABLE_LLVM=0 -DHOST_DEVICE_BUILD_HASH=00000000 -DENABLE_ICD=0 -DENABLE_LOADABLE_DRIVERS=0 -DENABLE_HOST_CPU_DEVICES=0 -DENABLE_HWLOC=0 -DENABLE_POCLCC=0 -DENABLE_TESTS=0 -DENABLE_EXAMPLES=0 -DBUILD_SHARED_LIBS=0 -DCMAKE_TOOLCHAIN_FILE=${ANDROID_NDK}/build/cmake/android.toolchain.cmake -DANDROID_NDK=${ANDROID_NDK} -DANDROID_PLATFORM=${ANDROID_PLATFORM_LEVEL} -DANDROID_ABI=${ANDROID_ABI} -DANDROID_NATIVE_API_LEVEL=${ANDROID_PLATFORM_LEVEL}

Could you provide the specific reasons why PoCL does not support CPU devices on Android?

From what I understand, it seems that LLVM libraries need to be statically linked for use on these devices, but the Android NDK does not offer such static libraries, making `LLVM.cmake` emit an error (llvm-config cannot find libraries). To address this, I compiled a specific LLVM version compatible with my NDK, and I am currently attempting to build with it.

However, before proceeding, I’d like to confirm if there are any additional challenges or limitations that I should be aware of. Can you explain why PoCL does not support Android with CPU devices?

**Comment by pjaaskel** [comment]
We have used PoCL on Android hosts very recently, but not with the online compiler support ([with the Proxy device and PoCL-Remote](https://www.youtube.com/playlist?list=PLS9btrvnAmoKR7D_O6S6zetc7np0wYDAK)). There are no fundamental obstacles that I know of. If you get LLVM built and linked with PoCL it should work easily. PoCL community will appreciate all possible fixes and documentation you can provide back related to your Android online compilation work.

Perhaps @RABijl can summarize his recent PoCL-on-Android experiences?

**Reply by widiba03304** [reply]
Thanks for the kind reply!

**Comment by RABijl** [comment]
I've been building PoCL for Android for a while now, and I think the biggest hurdle is compiling LLVM for Android. From the sound of it, you already did that. I'd be interested in learning how you did it. Besides that, modern Android versions really hammered down the file system and library loading. For example, you can not load any vendor libraries unless explicitly whitelisted. So there is a possibility that the compiled kernel file  (which is just a .so) might not be able to be loaded. Again, I have not tried this, but it is a possibility I can see happening. The fact that a lot of the first search results I see online for compiling on Android are to old projects and apps that have been pulled from the app store is not very encouraging.

If you do not require online compilation, you might get away with compiling them offline and including them in the apk, as I can confirm that loading .so files from there works. You could use `poclcc` to compile a kernel and then load it with `clCreateProgramWithBinary'. There might be a little bit of work required on the PoCL side to make this work since it has been a while since someone tried it.

**Reply by widiba03304** [reply]
Thanks for the reply! I built the LLVM cloned from the AOSP and checked out the commit hash that my NDK used (which can be known by running `clang --version`). I included ARM, AArch64, X86 and BPF targets because `llvm-config --libs --link-static` shows me it requires those. Additionally, llvm polly seems to be also required, but I have not checked thoroughly.

This is the cmake command that I used to build.
```
cmake \
    -DCMAKE_BUILD_TYPE="Release" \
    -DLLVM_ENABLE_PROJECTS=clang \
    -DCMAKE_INSTALL_PREFIX="$HOME/opt/llvm-14-android" \
    -DCMAKE_TOOLCHAIN_FILE="$ANDROID_NDK_HOME/build/cmake/android.toolchain.cmake" \
    -DANDROID_ABI="arm64-v8a" \
    -DLLVM_DEFAULT_TARGET_TRIPLE="aarch64-linux-android" \
    -DLLVM_TARGETS_TO_BUILD="ARM;AArch64;X86;BPF" \
    -DLLVM_TABLEGEN="../build/bin/llvm-tblgen" \
    -DCLANG_TABLEGEN="../build/bin/clang-tblgen" \
    -DLLVM_INCLUDE_BENCHMARKS=OFF \
    -DLLVM_INCLUDE_DOCS=OFF \
    -DLLVM_INCLUDE_EXAMPLES=OFF \
    -DLLVM_INCLUDE_RUNTIMES=OFF \
    -DLLVM_INCLUDE_TESTS=OFF \
    -DLLVM_INCLUDE_GO_TESTS=OFF \
    -DLLVM_INCLUDE_TOOLS=ON \
    -DLLVM_INCLUDE_UTILS=ON \
    -DLLVM_BUILD_BENCHMARKS=OFF \
    -DLLVM_BUILD_EXAMPLES=OFF \
    -DLLVM_BUILD_RUNTIME=OFF \
    -DLLVM_BUILD_RUNTIMES=OFF \
    -DLLVM_BUILD_TESTS=OFF \
    -DLLVM_BUILD_TOOLS=ON \
    -DLLVM_BUILD_UTILS=ON \
    -DLLVM_BUILD_LLVM_DYLIB=ON \
    -DLLVM_ENABLE_PIC=False \
    -G "Ninja" \
    ../llvm
```

I am not targeting the use of pocl in the app environment, but just in shell, so the whitelist is not a problem for me (I can just set LD_LIBRARY_PATH). My biggest concern now is to bypass `LLVM.cmake`. I will share the result after I try it. Thanks again!

**Reply by RABijl** [reply]
Thanks for sharing that config. As for building PoCL, have you tried adding `-DSTATIC_LLVM=ON` to the cmake arguments? And as for polly, I remember that also being required.

**Reply by RABijl** [reply]
there is also the `WITH_LLVM_CONFIG` option you can add to PoCL which is useful in scenarios where you have multiple LLVM installs.

**Reply by pjaaskel** [reply]
@franz see above, I believe you were working on a feature to allow bypassing LLVM.cmake in LLVM detection?

**Reply by widiba03304** [reply]
@RABijl Thanks for the guidance! From what I understand, `WITH_LLVM_CONFIG` sets up the executable that `LLVM.cmake` uses to list available libraries, targeting x86_64 architecture and adding those libraries to `LLVM_LIBFILES` for linking. Once configured, `LLVM_LIBFILES` is renamed and utilized in `lib/llvmopencl` and `lib/CL`, meaning the libraries need to be built for aarch64.

To address this, I'm aiming to bypass `LLVM.cmake` entirely to avoid using `llvm-config`. I’ll give it a try today or tomorrow and will update you if I run into any issues or make progress :)

@franz If you can give any advice, please teach me!

**Reply by franz** [reply]
@widiba03304  "targeting x86_64 architecture" - LLVM.cmake does not target any specific architecture, we use it with CUDA and also with ARM64 in our github CI. Do you mean you're trying to cross-compile PoCL for Android ? In that case yes, target is different than host, and LLVM.cmake would likely setup wrong values. If you managed to cross-compile LLVM for Android, and you have an Android shell, probably the simplest way to get a working PoCL is to copy the cross-compiled LLVM to the Android system, and build PoCL there directly. Is this an approach you cannot use for some reason ?

**Reply by widiba03304** [reply]
@franz Thanks for the suggestion! Yes, I am cross-compiling pocl. I'll check the way you suggested, but I think it is hard to build cmake, make, or ninja for Android. So, my attempt now is to override all CMake variables set by `LLVM.cmake` (`POCL_LLVM_LIBS`, `CLANG_LIBFILES` and `LLVM_SYSLIBS` maybe?).

**Reply by widiba03304** [reply]
BTW, I also set `LLVM_OPT` `LLVM_LLC`, `LLVM_AS`, `LLVM_LINK`, and `LLVM_LLI` with x86_64 binaries as `sleef` kernels will be compiled on the host. Additionally, I found llvm-14, which is the version that my NDK uses, has a bug making `llvm-link` crash while compiling `sleef` (I think this happens only when targeting `LLC_TRIPLE=aarch64-linux-android`, but I didn't check thoroughly.), so I upgraded the LLVM version of the host to 19.

**Comment by widiba03304** [comment, marked as answer]
After days of effort, I finally managed to compile and run the pocl on Android, with `LLVM.cmake` proving to be particularly challenging. Here are some key changes I made to get everything working:

1. The current build script only supports the host machine, which complicates cross-compiling. To fix this, I modified `config.h.in.cmake` to recognize cross-compilation variables like `CLANG`, `CLANGXX`, `LLVM_LLC`, `LLVM_OPT`, etc. for the target machine. I also introduced additional CMake variables (CLANG_BIN_PATH, CLANGXX_BIN_PATH, etc.) to ensure it doesn’t default to the host’s paths.
2. By default, DEFAULT_HOST_LD_FLAGS points to `/system/lib/crtbegin_so.o` and `/system/lib/crtend_so.o`, which aren’t typically installed on Android devices. Since these files are included in the NDK, I added steps to push them to the device.
3. The `compile_c_to_bc` function in `bitcode_rules.cmake` doesn’t account for the NDK’s sysroot by default. To address this, I manually added the --sysroot flag.

There were also several minor adjustments I made along the way. They’re relatively straightforward to fix but too numerous to list here.

Thanks to everyone for the inputs and advice! @pjaaskel @RABijl @franz

**Reply by pjaaskel** [reply]
Great to hear! Would you like to contribute your Android improvements (and documentation) back to the PoCL upstream project to benefit the community and help others improve the Android port even further?

**Reply by widiba03304** [reply]
Thank you so much for the suggestion! I’d really love to participate. For now, though, I need to focus on finishing my current project, so I’ll take a little time to organize my schedule first. I’ll reach out when I’m ready.
